Move spec tests into the LLB graph

cpuguy83 opened the pull request:
This moves Dalec spec/package tests into the normal LLB graph instead of running test containers manually through the BuildKit API after the main solve.

The main implementation details and rationale are in `tests: run package tests in LLB graph`.

This also includes small supporting fixes to protect shared test spec mutation and to build docs examples through the existing bake target.

cpuguy83 committed:
ci: build examples through bake

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
tests: protect shared spec defaults

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
tests: run package tests in LLB graph

Move spec/package test execution out of the frontend's manual container-run path and into the normal LLB graph.

The previous test runner solved the package build, then used the BuildKit API to run test containers and inspect filesystem state from outside the graph. That made the build less lazy, introduced extra solve behavior, and broke clients such as `docker buildx dap build` that expect the frontend solve to remain debuggable as a normal graph.

This change represents test execution as BuildKit operations instead. The frontend binary is mounted into test containers and invoked through internal subcommands for filesystem checks. Each assertion is emitted as a separate exec op so failures can retain useful source mapping.

Running tests this way keeps the test work attached to the build graph, lets BuildKit schedule and cache it normally, and avoids forcing package outputs to be materialized just so the frontend can run tests manually.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented on tests.go:
```diff
@@ -61,26 +58,38 @@ type TestStep struct {
 
-// GetSourceLocation returns an llb.ConstraintsOpt for the TestStep
+// GetSourceLocation returns the source mapping for the TestStep's command
 func (ts *TestStep) GetSourceLocation(state llb.State) llb.ConstraintsOpt {
-	return ts._sourceMap.GetLocation(state)
+	return ts._commandSourceMap.GetLocation(state)
 }
```
nil is handled on sourcemap.

cpuguy83 commented on tests.go:
```diff
@@ -413,118 +423,38 @@ func (c *FileCheckOutput) processBuildArgs(lex *shell.Lex, args map[string]strin
-		for i, v := range c.Contains {
-			if v == err.Expected && i < len(c.containsSourceMaps) && c.containsSourceMaps[i] != nil {
-				return c.containsSourceMaps[i].GetErrdefsSource()
-			}
-		}
+		return c.containsSourceMaps[index].GetLocation(state)
 	case CheckOutputMatchesKind:
-		for i, v := range c.Matches {
-			if v == err.Expected && i < len(c.matchesSourceMaps) && c.matchesSourceMaps[i] != nil {
-				return c.matchesSourceMaps[i].GetErrdefsSource()
-			}
-		}
+		return c.matchesSourceMaps[index].GetLocation(state)
 	case CheckOutputStartsWithKind:
```
nil is handled on sourcemap.

cpuguy83 commented on cmd/frontend/main.go:
```diff
@@ -28,43 +31,62 @@ func init() {
+
+	if h == nil {
+		fmt.Fprintln(os.Stderr, "unknown subcommand:", flags.Arg(1))
+		fmt.Fprintln(os.Stderr, "full args:", flag.Args())
```
Fixed.

cpuguy83 commented on internal/testrunner/runner.go:
```diff
@@ -0,0 +1,271 @@
+			i.ExtraOpts = append(i.ExtraOpts, mnt.ToRunOption(sOpt, dalec.WithConstraints(opts...)))
+		}
+	}
+}
```
Fixed.

cpuguy83 commented on internal/testrunner/runner.go:
```diff
@@ -0,0 +1,271 @@
+	return p
+}
+
+// WithFinalState returns a state option which takes as input a potentially modified
+// state and returns the original unmodified state.
+// This makes sure that any changes made during test steps are discarded but makes sure
+// there is a dependency on the intermediate state so buildkit will execute it.
+//
+// NOTE: This is a hack to work around the fact that buildkit does not currently
+// have a proper way to express "run this for validation only".
+func WithFinalState(st llb.State, opts ...ValidationOpt) llb.StateOption {
+	return trueCmd.WithOutput(st, opts...)
+}
```
Incorrect.

cpuguy83 commented on tests.go:
```diff
@@ -413,118 +423,38 @@ func (c *FileCheckOutput) processBuildArgs(lex *shell.Lex, args map[string]strin
+	switch kind {
 	case CheckFilePermissionsKind:
-		return c.permissionsSourceMap.GetErrdefsSource()
+		return c.permissionsSourceMap.GetLocation(state)
 	case CheckFileIsDirKind:
-		return c.isDirSourceMap.GetErrdefsSource()
+		return c.isDirSourceMap.GetLocation(state)
 	case CheckFileNotExistsKind:
-		return c.notExistSourceMap.GetErrdefsSource()
+		return c.notExistSourceMap.GetLocation(state)
 	case CheckFileLinkTargetPathKind:
-		return c.linkTargetSourceMap.GetErrdefsSource()
+		return c.linkTargetSourceMap.GetLocation(state)
 	default:
 		// Delegate to embedded CheckOutput (equals/contains/...)
-		return c.CheckOutput.GetErrSource(err)
+		return c.CheckOutput.GetSourceLocation(state, kind, index)
 	}
```
nil is handled.

cpuguy83 commented on frontend/build.go:
```diff
@@ -54,7 +54,7 @@ func LoadSpec(ctx context.Context, client *dockerui.Client, platform *ocispecs.P
 	}
 
-	src, err := client.ReadEntrypoint(ctx, "Dockerfile")
+	src, err := client.ReadEntrypoint(ctx, "dalec")
```
This is the language, not a filename and is used for sourcemapping.

MorrisLaw commented on internal/testrunner/file_startswith.go:
```diff
@@ -0,0 +1,66 @@
+type checkFileStartsWithCommand string
+
+func (c checkFileStartsWithCommand) WithCheck(p string, checker *dalec.CheckOutput, opts ...ValidationOpt) llb.StateOption {
+	return func(in llb.State) llb.State {
```
Should we also have an empty check here?

```go
if checker.StartsWith == "" {
    return in
}
```

cpuguy83 replied:
Done.

MorrisLaw commented on internal/testrunner/file_empty.go:
```diff
@@ -0,0 +1,74 @@
+
+	var actual string
+
+	f, err := mmapFile(p)
```
should we `defer f.close()`?

cpuguy83 replied:
Not really neccessary since we are exiting the process almost immediately, but added anyway.

MorrisLaw commented on targets/linux/deb/distro/pkg.go:
```diff
@@ -178,11 +172,12 @@ func addPaths(paths []string, opts ...llb.ConstraintsOpt) llb.StateOption {
-		return in.With(deps).With(tests)
+		opts = append(opts, frontend.IgnoreCache(client))
+		withTestDeps := cfg.InstallTestDeps(sOpt, targetKey, spec, opts...)
+		runTests := frontend.RunTests(ctx, client, sOpt, spec, withTestDeps, targetKey)
```
Can we pass in `opts...` here like we do in the rpm version? The constraints get passed through to the testrunner.WithConstraints call in RunTests; omitting it might lose out on some constraints for the deb tests otherwise.

cpuguy83 replied:
Nice catch, fixed.

MorrisLaw commented on targets/linux/rpm/distro/pkg.go:
```diff
@@ -43,17 +45,11 @@ func needsAutoGocache(spec *dalec.Spec, targetKey string) bool {
 
-	// Preprocess the spec to generate patches for gomod edits and other generators
-	// This must happen after build deps are installed so Go is available
-	if err := spec.Preprocess(sOpt, worker, opts...); err != nil {
```
Did we intend to remove the calls to `spec.Preprocess(sOpt, worker, opts...)` for both rpm and deb? Has it been replaced with something else?

cpuguy83 replied:
Looks like bad rebase removed this. *Really* good catch, though I'm a little disheartened that this isn't breaking a test.

cpuguy83 replied:
Added a proper test case for this.
There was already one but not testing at the distro-level which itself is responsible for handling the preprocess step.

cpuguy83 committed:
Add distro test for go mod replace

This is important to test at the distro level since the distro has to
add the call to actually handle this.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>
